### PR TITLE
Discovery refactor

### PR DESCRIPTION
**Pull request by crazybarber**
This refactor aims to split two scope abstracts: local and global scope. The global scope is responsible for interactive installation of Klio modules. The local scope is responsible for installing all dependencies as defined in project definition yaml. Previously those two scopes were if-ed through get command body. Now both are split up covered by a shared interface. 

Such an operation also required minor refactor in Path validation and discovery. More details in dedicated comments.

### CODE REVIEW

File: internal/context/context.go [hunk 1]
```diff
@@ -1,8 +1,9 @@
 package context
 
 type CLIContext struct {
-	Config CLIConfig
-	Paths  Paths
+	Config              CLIConfig
+	Paths               Paths
+	ProjectConfigExists bool
```
Comment (crazybarber) [inline internal/context/context.go:6]: This is an addition in order to mark (on discovery) if the project config yaml exists.

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Review by kstasik — APPROVED** [review]

File: internal/cmd/get/get.go [hunk 2]
```diff
@@ -45,128 +42,25 @@ func NewCommand(ctx context.CLIContext) *cobra.Command {
...
-		if projectConfig.DefaultRegistry != "" {
-			depsMgr.DefaultRegistry = projectConfig.DefaultRegistry
-		}
-	}
-
-	var toInstall []schema.Dependency
-
-	if len(args) == 0 && !opts.Global {
-		toInstall = projectConfig.Dependencies
+		getScope = scope.NewGlobal(ctx.Paths.GlobalInstallDir, opts.From, opts.As, opts.Version)
```
Comment (qzb) [inline internal/cmd/get/get.go:48]: Options *from*, *as* and *version* are describing command which is going to be installed, not scope. These options apply also to commands installed locally.

File: internal/cmd/get/get.go [hunk 3]
```diff
@@ -45,128 +42,25 @@ func NewCommand(ctx context.CLIContext) *cobra.Command {
...
-	err := os.MkdirAll(ctx.Paths.ProjectInstallDir, 0o755)
+	err = getScope.Initialize(&ctx)
 	if err != nil {
-		return ctx, err
+		log.Fatalf("scope initialization failed: %s", err)
 	}
-
-	// create and save default klio config
-	_, err = schema.CreateDefaultProjectConfig(ctx.Paths.ProjectConfigFile)
+	err = getScope.InstallDependencies(args)
```
Comment (qzb) [inline internal/cmd/get/get.go:61]: This is confusing. It looks like this installs commands from args, but in reality:
- `args` cannot have more than 1 element
- for local scope `args` may be empty
- if `args` have 0 elements, actual list of dependencies is taken from project config.

In my opinion you went too far with removing "ifs". Simply add additional step for obtaining list of dependencies to install:

```go
var deps []schema.Dependency

switch len(args) {
case 0:
	deps = scope.GetDependencies() // for global scope this should always return an empty list
case 1:
	deps = []schema.Dependency{
 		{
 			Name:     args[0],
 			Version:  opts.Version,
 			Registry: opts.From,
 			Alias:    opts.As,
 		},
	}
default:
	log.Fatal("...")
}

if len(deps) == 0 {
	log.Fatal("...")
}

err = getScope.InstallDependencies(deps)
```

File: internal/cmd/get/get.go [hunk 4]
```diff
@@ -45,128 +42,25 @@ func NewCommand(ctx context.CLIContext) *cobra.Command {
...
 	if err != nil {
-		return ctx, err
+		log.Fatalf("installing dependencies failed: %s", err)
 	}
-
-	return ctx, nil
+	log.Info(getScope.GetSuccessMsg())
```
Comment (qzb) [inline internal/cmd/get/get.go:65]: I'm not sure, if scope should be responsible for generating this success message. There is no need for separate messages for local and global scope.

File: internal/scope/global.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		},
+	}
+
+	_ = installDependencies(g.dependencyManager, dep, dependency.GlobalScope)
+
+	return nil
+}
+
+func (g *global) GetSuccessMsg() string {
+	return fmt.Sprintf("command %s:%s installed successfully", g.commandName, g.CommandVersionOption)
```
Comment (qzb) [inline internal/scope/global.go:66]: If this function needs commandName, it should receive it through parameters.

**Review by qzb — CHANGES_REQUESTED** [review]
If we had some tests, they would fail 😄 Generally I like this changes a lot, but you need polish some parts and *test it*.

File: internal/context/paths.go [hunk 6]
```diff
@@ -1,77 +1,62 @@
...
-	if result.ProjectConfigFile != "" {
-		result.ProjectInstallDir = path.Join(filepath.Dir(result.ProjectConfigFile), cfg.InstallDirName)
+	workDir, err := getWorkDirPath()
+	if err != nil {
+		return Paths{}, err
 	}
 
-	return result
+	return Paths{
+		ProjectConfigFile: path.Join(workDir, cfg.ProjectConfigFileName),
```
[the remark this answers is not in the document]
Comment (crazybarber) [reply internal/context/paths.go]: Please review. Whole path validation done now on Init

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

File: internal/cmd/get/get.go (hunk 3 above)
Comment (crazybarber) [reply internal/cmd/get/get.go:61]: `type Dependency struct` + `package dependency` = good package
Code: `[]schema.Dependency` >.<

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

File: internal/scope/local.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		if err := schema.SaveProjectConfig(l.projectConfig); err != nil {
+			return fmt.Errorf("unable to update dependencies in the %s file: %s", l.ProjectConfigFile, err)
+		}
+	}
+
+	return nil
+}
+
+func (l *local) GetInstalledDependencies() []dependency.Dependency {
+	return l.projectConfig.Dependencies
```
Comment (qzb) [inline internal/scope/local.go:119]: Shouldn't you use `l.installedDeps`?

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Review by qzb — APPROVED** [review]

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Comment by g2a-official** [comment]
**Coverage report**
github.com/g2a-com/klio/internal/cmd/get/get.go:26:	NewCommand			0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:47:	run				0.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:143:	initialiseProjectInCurrentDir	75.0%
github.com/g2a-com/klio/internal/cmd/get/get.go:154:	initialiseProject		77.8%
github.com/g2a-com/klio/internal/schema/schema.go:57:	NewDefaultProjectConfig		100.0%
github.com/g2a-com/klio/internal/schema/schema.go:65:	MarshalYAML			60.0%
github.com/g2a-com/klio/internal/schema/schema.go:124:	minimalKlioFile			100.0%
github.com/g2a-com/klio/internal/schema/schema.go:142:	UnmarshalYAML			54.5%
github.com/g2a-com/klio/internal/schema/util.go:17:	LoadConfigFile			0.0%
github.com/g2a-com/klio/internal/schema/util.go:55:	SaveConfigFile			58.8%
github.com/g2a-com/klio/internal/schema/util.go:88:	LoadProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:97:	SaveProjectConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:102:	LoadCommandConfig		0.0%
github.com/g2a-com/klio/internal/schema/util.go:111:	LoadDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:120:	SaveDependenciesIndex		0.0%
github.com/g2a-com/klio/internal/schema/util.go:125:	CreateDefaultProjectConfig	83.3%
github.com/g2a-com/klio/internal/schema/util.go:151:	validate			57.1%
total:							(statements)			35.5%

**Pull request merged**